Userland/Libraries/LibGUI/BoxLayout.cpp

```diff
@@ -23,74 +24,118 @@ BoxLayout::BoxLayout(Orientation orientation)
+        return { result_primary, result_secondary };
     else
-        size += margins().top() + margins().bottom() + content_margins.top() + content_margins.bottom();
-
-    if (!size)
-        return -1;
-    return size;
+        return { result_secondary, result_primary };
 }
 
-int BoxLayout::preferred_secondary_size() const
+UISize BoxLayout::min_size() const
 {
-    auto widget = verify_cast<GUI::Widget>(parent());
-    int size = 0;
+    VERIFY(m_owner);
+
+    UIDimension result_primary { 0 };
+    UIDimension result_secondary { 0 };
+
+    bool first_item { true };
     for (auto& entry : m_entries) {
         if (!entry.widget || !entry.widget->is_visible())
             continue;
-        int min_size = entry.widget->min_size().secondary_size_for_orientation(orientation());
-        int preferred_secondary_size = -1;
-        if (entry.widget->is_shrink_to_fit() && entry.widget->layout()) {
-            preferred_secondary_size = entry.widget->layout()->preferred_size().secondary_size_for_orientation(orientation());
-            size = max(size, preferred_secondary_size);
+
+        UISize min_size = entry.widget->effective_min_size();
+
+        {
+            UIDimension primary_min_size = min_size.primary_size_for_orientation(orientation());
+
+            VERIFY(primary_min_size.is_one_of({ GUI::SpecialDimension::Shrink, GUI::SpecialDimension::Regular }));
+
+            if (primary_min_size.is_regular_value())
+                result_primary.add_if_regular(primary_min_size.value_verify_regular());
+
+            if (!first_item)
+                result_primary.add_if_regular(spacing());
+        }
+
+        {
+            UIDimension secondary_min_size = min_size.secondary_size_for_orientation(orientation());
+
+            VERIFY(secondary_min_size.is_one_of({ GUI::SpecialDimension::Shrink, GUI::SpecialDimension::Regular }));
+
+            result_secondary = max(result_secondary, secondary_min_size);
         }
-        size = max(min_size, size);
+
+        first_item = false;
     }
 
-    auto content_margins = widget->content_margins();
+    result_primary.add_if_regular(
+        margins().primary_total_for_orientation(orientation())
+        + m_owner->content_margins().primary_total_for_orientation(orientation()));
+
+    result_secondary.add_if_regular(
+        margins().secondary_total_for_orientation(orientation())
+        + m_owner->content_margins().secondary_total_for_orientation(orientation()));
+
     if (orientation() == Gfx::Orientation::Horizontal)
-        size += margins().top() + margins().bottom() + content_margins.top() + content_margins.bottom();
+        return { result_primary, result_secondary };
     else
-        size += margins().left() + margins().right() + content_margins.left() + content_margins.right();
-
-    if (!size)
-        return -1;
-    return size;
+        return { result_secondary, result_primary };
 }
 
 void BoxLayout::run(Widget& widget)
```

Comment:
This is a big enough change (change the algorithm used to run box layout), figure out the minimal changes to fix the old one when you change the BoxLayout api, then later change it

Reply:
That won't work. The old one was fine, it just didn't have all the features. And I can't add the features if I don't have the new information.
I would have to implement the function entirely new, just to throw it away one commit later. That seems nonsensical.

Reply:
That can be useful, you can deprecate the current api for BoxLayout, add the planned api to BoxLayout, then do a everywhere commit to upgrade all the uses of the cur api to the planned one, then finally remove the no longer useful api

Reply:
And inbetween when both the old and new work, add a commit that changes to the new BoxLayout::run or maybe the commit focused on only BoxLayout is no longer massive